### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by sheshtawy**

### CODE REVIEW

File: quart_prometheus_logger.py [hunk 1]
```diff
@@ -66,14 +68,15 @@ def __init__(self, app: Optional[Quart] = None, metrics_endpoint: str = "root"):
...
+        self._registry = CollectorRegistry()
         self._register_collectors()
         if app:
             self.init_app(app, metrics_endpoint)
 
     def init_app(self, app: Quart, metrics_endpoint: str):
         """Register an application."""
 
-        def start_request():
+        def prometheus_before_request_callback():
```
Comment (arusahni) [inline quart_prometheus_logger.py]: These feel like a bit of smurf naming. Maybe event-handler naming is more explanatory? e.g., `on_request_start()`, `on_request_end()`, `on_request_error()`?

File: quart_prometheus_logger.py [hunk 2]
```diff
@@ -105,17 +108,27 @@ def end_request(response):
...
-            return end_request(response)
+            return prometheus_after_request_callback(response)
 
         self.app = app
-        app.before_request(start_request)
-        app.after_request(end_request)
-        app.register_error_handler(HTTPStatusException, abort_with_error)
+        app.before_request(
+            prometheus_before_request_callback,
+            prometheus_before_request_callback.__name__,
```
Comment (arusahni) [inline quart_prometheus_logger.py]: This parameter is not the name of the function (despite being called `name`). It is the name of the blueprint to which the handler is attached. Since this isn't attached to a specific blueprint, we shouldn't be providing a key.

File: quart_prometheus_logger.py [hunk 3]
```diff
@@ -105,17 +108,27 @@ def end_request(response):
...
+        )
+        app.after_request(
+            prometheus_after_request_callback,
+            prometheus_after_request_callback.__name__,
+        )
+        app.register_error_handler(
+            HTTPStatusException,
+            prometheus_error_handler,  # type:ignore
```
Comment (arusahni) [inline quart_prometheus_logger.py]: What error are we suppressing here? If possible, we should correct the type mismatch.

File: quart_prometheus_logger.py [hunk 4]
```diff
@@ -66,14 +68,15 @@ def __init__(self, app: Optional[Quart] = None, metrics_endpoint: str = "root"):
         self._collectors: Dict[str, MetricType] = {}
         self._custom_labeler: Optional[Callable[["LocalProxy"], Dict[str, str]]] = None
         self._custom_label_names: List[str] = []
+        self._registry = CollectorRegistry()
```
Comment (arusahni) [inline quart_prometheus_logger.py:74]: I'm fine with no longer using an explicitly-managed registry, but we should ensure this is identically instantiated. As of now, it is missing the `auto_describe` parameter.

Additionally, if folks want to manage it externally, I think it'd be useful to add this as an optional argument.

File: quart_prometheus_logger.py [hunk 5]
```diff
@@ -127,46 +140,57 @@ def _register_collectors(self):
...
                     buckets=RESPONSE_BUCKETS,
+                    registry=self._registry,
                 ),
             )
         }
 
+    def _reset_metrics(self):
+        """Unregister all collectors from the registry."""
+        for _, collector in self._collectors.items():
+            self._registry.unregister(collector)
```
Comment (arusahni) [inline quart_prometheus_logger.py:180]: Not a critique of this PR, just annoyed that they didn't go with `deregister` :disappointed:

File: quart_prometheus_logger.py [hunk 6]
```diff
@@ -127,46 +140,57 @@ def _register_collectors(self):
...
+
     def custom_route_labeler(
         self, labeler: Callable[["LocalProxy"], Dict[str, str]], label_names: List[str]
     ) -> None:
         """Add a handler for providing additional labels for a route.
 
         This will reset all metrics. Conventionally it's called when the extension is first registered
 
-        :param labeler: The handler function to invoke. It must return a dict of key-value labels.
+        :param labeler: The handler function to invoke. It takes a Quart LocalProxy representing
```
Comment (arusahni) [inline quart_prometheus_logger.py]: I feel like the "Quart LocalProxy" part is implied by the type signature. We should be able to rephrase this to something like: "It must accept a request proxy".

File: quart_prometheus_logger.py [hunk 7]
```diff
@@ -127,46 +140,57 @@ def _register_collectors(self):
...
+                    the request, from which values for the custom label can be pulled. It must return
```
Comment (arusahni) [inline quart_prometheus_logger.py]: Instead of "pulled", maybe something a little more active like "constructed" or "generated"?

File: setup.py [hunk 8]
```diff
@@ -1,50 +1,47 @@
...
-        'Operating System :: OS Independent',
-        'Programming Language :: Python',
-        'Topic :: Internet :: WWW/HTTP :: Dynamic Content',
-        'Topic :: Software Development :: Libraries :: Python Modules'
-    ]
-)
\ No newline at end of file
+        "Environment :: Web Environment",
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: BSD License",
```
Comment (arusahni) [inline setup.py]: License mismatch? It should be MIT.

File: tests/test_quart_prometheus_logger.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import pytest
+import pytest_asyncio
+import quart
+
+from quart_prometheus_logger import PrometheusRegistry
+from prometheus_client import REGISTRY
```
Comment (arusahni) [inline tests/test_quart_prometheus_logger.py:6]: Unused import

File: tests/test_quart_prometheus_logger.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    assert "prometheus_after_request_callback" in app.after_request_funcs
+
+
+@pytest.mark.asyncio
+async def test_customer_labeler(app, client):
+    registry = PrometheusRegistry(app=app)
+    custom_label = lambda _: {"custom_label": 1}
+    registry.custom_route_labeler(custom_label, ["custom_label"])
+    await client.get("/")
+    for _, collector in registry._collectors.items():
```
Comment (arusahni) [inline tests/test_quart_prometheus_logger.py:31]: Nice

**Review by arusahni — CHANGES_REQUESTED** [review]
One more request (can't make it using the review tool because it hasn't been changed in this PR): Could we change the `metrics_endpoint`'s default value to `None` (making it an `Optional[str]`? The current default reflects our internal convention (vs. something meaningful to Quart). [Quart's default is `None`](https://pgjones.gitlab.io/quart/reference/source/quart.app.html?highlight=add_url_rule#quart.app.Quart.add_url_rule), so we should be able to punch that through.

File: quart_prometheus_logger.py (hunk 4 above)
Comment (sheshtawy) [reply quart_prometheus_logger.py:74]: that's a good idea

File: quart_prometheus_logger.py (hunk 2 above)
Comment (sheshtawy) [reply quart_prometheus_logger.py]: I understand. The reason I added a name, was to be able to test it and find it in the the before_request handlers, if I remove it, maybe I can adjust the unit tests to assert the count of handlers before and after registration?

File: quart_prometheus_logger.py (hunk 1 above)
Comment (sheshtawy) [reply quart_prometheus_logger.py]: I like that more 👍

File: quart_prometheus_logger.py (hunk 3 above)
Comment (sheshtawy) [reply quart_prometheus_logger.py]: There is a typing mismatch between what Quart expects and what our handler actually does. Even in Quart's example use of the function there will be a type error if mypy is run over the example. https://pgjones.gitlab.io/quart/reference/source/quart.app.html#quart.app.Quart.register_error_handler

File: quart_prometheus_logger.py (hunk 5 above)
Comment (sheshtawy) [reply quart_prometheus_logger.py:180]: 😞
